### ISSUE TITLE
Specifying subnet-ids when creating a new cluster with private topology breaks install

### ISSUE THREAD

**Issue by geojaz** [kubernetes/kops#1147]
As per [Docs: Run in Existing VPC](https://github.com/kubernetes/kops/blob/master/docs/run_in_existing_vpc.md), I am using `kops`  to spin up a new cluster in an existing VPC that has public and private subnets. I am using the private topology mode and my goal is to input existing subnet-ids so that `kops`  can put the public facing resources (ELBs for api and bastion) inside.

I start my cluster creation with the following command:

```yaml
    kops create cluster \
      --cloud aws \
      --topology private \
      --networking weave \
      --state $KOPS_STATE_STORE \
      --node-count $NODE_COUNT \
      --zones $ZONES \
      --master-zones $MASTER_ZONES \
      --dns-zone $DNS_ZONE \
      --node-size $NODE_SIZE \
      --master-size $MASTER_SIZE \
      -v $V_LOG_LEVEL \
      --ssh-public-key $SSH_KEY_PATH \
      --vpc $VPC_ID \
      --network-cidr $NETWORK_CIDR \
      --name=$CLUSTER_NAME
```
When I edit the cluster with `kops edit cluster $CLUSTER_NAME` , I update the zones section as follows. For clarity, the CIDRS 172.20.*/21 that I’m specifying are unused blocks that I want `kops` to allocate for my private subnets. I then specify the IDs of the public subnets that I wish to reuse for my ELBs.
```yaml
      zones:
      - cidr: 172.20.120.0/21
        name: us-east-1a
        id: subnet-1111
      - cidr: 172.20.128.0/21
        name: us-east-1b
        id: subnet-2222
      - cidr: 172.20.136.0/21
        name: us-east-1c
        id: subnet-3333
```
`kops edit cluster $CLUSTER_NAME` parses these changes successfully, but when I run `kops update cluster $CLUSTER_NAME` , it stops processing while building the tasks list here:

`error building tasks: unexpected error resolving task "autoscalingGroup/nodes.erichole.dev.example.com": Unable to find task "subnet/private-us-east-1a.erichole.dev.example.com", referenced from autoscalingGroup/nodes.erichole.dev.example.com:.Subnets[0]`

**Comment by caseeker** [comment]
The work around I'm using for now is simply leaving the topology as public.

**Comment by geojaz** [comment]
That does make it a little bit easier. Unfortunately we require private topology for our use-case. :)

In any event, as per @justinsb it seems there's a bug in the SharedVPC conditional in https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/_aws/topologies/_topology_private/network.yaml . I'm working on it, first to try get the word out that this feature exists via some documentation updates- and then to squash this bug.

**Comment by kris-nova** [comment]
@geojaz 

Can we get some more detail on this bug? The yaml blocks point to each other, and typically generate each other's ID's programmatically 

I am assuming there has to be something going on with the `id` directive here

```
      - cidr: 172.20.136.0/21
        name: us-east-1c
        id: subnet-3333
```

Happy to help bang on this with you if you need an extra pair of eyes - LMK

**Comment by geojaz** [comment]
Hey @kris-nova! 
In my understanding, the ID should be generated programatically when `kops` is building all the subnets from scratch. But according to the docs and @justinsb, `kops` can also re-use subnets if you input these IDs. So I've been experimenting-

I can get `kops` to "share" the public  subnets with this syntax:
``` 
      - cidr: 172.20.136.0/21
        name: us-east-1
        id: subnet-3333
```

But it breaks when I try to specify subnet ids to share in private mode. It dies during the generation of tasks in `kops update cluster $CLUSTER_NAME` with the error about building tasks above. Justin seemed to think this was just a problem with the `{{ if not (SharedZone $zone) }}` block, [specifically this](https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/_aws/topologies/_topology_private/network.yaml#L133-L143)? He reasoned that we'd still want to create those private subnets in a shared VPC. I've tried pulling the private subnet declaration out of the `{{ if not (SharedZone $zone) }}` block and update cluster gets out of the `loader` and into `executor` but now it dies because `kops` tries to change the CIDR fields on all the private networks. This doesn't make any sense to me because I want all new private subnets, so `kops` is seems to be misunderstanding my intentions.

There very well may be a bug here, but based on this experimentation, I think we may be missing something as well. It looks to me like this code to launch k8s in an existing VPC with shared subnets is really only set up for public networking? Further, I think my subnet specification syntax above is ambiguous in private mode. In order to share subnets in public networking, you specify both the CIDR of the subnet AND the subnet id. For private networking, I think you should provide the same information for public and private subnets.

After thinking about this a bunch and looking around the code, I think we may be in a "part bug, part feature" situation. I'm hopeful, but not sure that the work in #1110 will address these concerns. I think we should probably work to allow specification of CIDRs and optional subnet ids explicitly for both public and private topologies.  Perhaps you have some thoughts based on what I'm seeing.

**Comment by chrislovecnm** [comment]
We have PR inbound thay may help with this ... I am looking through issues to test against https://github.com/kubernetes/kops/pull/1183 - no promises but we need to test.

**Comment by chrislovecnm** [comment]
See https://github.com/kubernetes/kops/issues/1206 ~ The PR is in.  Who can test?

**Comment by geojaz** [comment]
I've tested this PR and provided feedback to @justinsb. It seems to be moving in the right direction, but not yet working. My flow is:
* Create a new cluster in existing VPC with `kops create cluster --topology private ...`
* Edit my cluster with `kops edit cluster $CLUSTER_NAME` and update the subnet section as follows:
```
 subnets:
  - cidr: 172.20.120.0/21
    name: us-east-1a
    type: Private
    zone: us-east-1a
  - cidr: 172.20.32.0/21
    name: utility-us-east-1a
    id: subnet-123456
    type: Utility
    zone: us-east-1a
```

I am able to complete the `kops update cluster` dry-run, but the output creates some concern. It shows that the ids for the utility subnets are not being saved after the edit (only the CIDRs are saved). If I run `kops edit` and input as above, then save my changes, then `kops edit` again, all my id fields vanish, although all my other changes stick around. Justin explained to me that currently the v1alpha2 code is being translated back into v1alpha1 code to write to disk for the moment until we're comfortable to go forward with v1alpha2 on disk. And v1alpha1 doesn't have a notion of custom subnet ids for utility subnets. So it loses my updates and can't apply them. 

This update is rad and I can't wait to see it all come together! Please let me know if I can provide any other info that would help get this one nailed down!

**Comment by rokka-n** [comment]
Sorry, which version of kops has `topology` option?

```
➜  terraform git:(kops) ✗ kops version
Version 1.4.4
➜  terraform git:(kops) ✗ kops create cluster --cloud=aws --topology private --vpc=vpc-xxxxxxx\
  --dns-zone=domain.com --master-size=t2.medium  \
   --master-zones us-west-2a \
  --network-cidr=10.5.0.0/16 --node-count=3 \
  --node-size=m3.medium \
  --zones=us-west-2a,us-west-2b,us-west-2c \
  --name=us-west-2.k8s.domain.com
...
unknown flag: --topology
```

**Comment by geojaz** [comment]
@rokka-n Download and compile `kops v1.5.0-alpha3`. Or from the `master` branch. Good luck!

**Comment by geojaz** [comment]
This is closed via `kops v1.5-alpha`. :D
